lib/SILGen/SILGenExpr.cpp

```diff
@@ -2149,11 +2150,109 @@ class NominalTypeMemberRefRValueEmitter {
 
 } // end anonymous namespace
 
+// We may simply forward the Optional<SomeClass> value as SomeClass if the
+// receiver of the value is the self argument of objc_msgSend.
+BindOptionalExpr *SILGenFunction::maySkipSwitchInBindOptionalSubExpr(Expr *E) {
```

Comment:
All the checks here are duplicating existing logic elsewhere in SILGen that tries to determine if a method is dispatched via objc_msgSend(). This is going to be a source of latent bugs.